common/modules/farm/farm_base_impl/src/exit_farm.rs

```diff
@@ -26,67 +25,62 @@ pub trait BaseExitFarmModule:
     + crate::base_farm_validation::BaseFarmValidationModule
-    + crate::partial_positions::PartialPositionsModule
+    + utils::UtilsModule
 {
-    fn exit_farm_base<
-        AttributesType,
-        GenerateAggregattedRewardsFunction,
-        CalculateRewardsFunction,
-    >(
+    fn exit_farm_base<FC: FarmContract<FarmSc = Self>>(
         &self,
+        caller: ManagedAddress,
         payment: EsdtTokenPayment<Self::Api>,
-        generate_rewards_fn: GenerateAggregattedRewardsFunction,
-        calculate_rewards_fn: CalculateRewardsFunction,
-    ) -> InternalExitFarmResult<Self, AttributesType>
-    where
-        AttributesType: Clone
-            + TopEncode
-            + TopDecode
-            + NestedEncode
-            + NestedDecode
-            + CurrentFarmAmountGetter<Self::Api>
-            + InitialFarmingAmountGetter<Self::Api>
-            + CompoundedRewardAmountGetter<Self::Api>,
-        GenerateAggregattedRewardsFunction: Fn(&Self, &mut StorageCache<Self>),
-        CalculateRewardsFunction:
-            Fn(&Self, &BigUint, &AttributesType, &StorageCache<Self>) -> BigUint,
-    {
+    ) -> InternalExitFarmResult<Self, FC::AttributesType> {
         let mut storage_cache = StorageCache::new(self);
-        let exit_farm_context =
-            ExitFarmContext::new(payment, &storage_cache.farm_token_id, self.blockchain());
+        let exit_farm_context = ExitFarmContext::<Self::Api, FC::AttributesType>::new(
+            payment,
+            &storage_cache.farm_token_id,
+            self.blockchain(),
+        );
 
         self.validate_contract_state(storage_cache.contract_state, &storage_cache.farm_token_id);
-        generate_rewards_fn(self, &mut storage_cache);
+        FC::generate_aggregated_rewards(self, &mut storage_cache);
 
         let farm_token_amount = &exit_farm_context.farm_token.payment.amount;
-        let attributes = &exit_farm_context.farm_token.attributes;
-        let mut reward = calculate_rewards_fn(self, farm_token_amount, attributes, &storage_cache);
-        storage_cache.reward_reserve -= &reward;
-
-        let prev_compounded_rewards =
-            self.calculate_previously_compounded_rewards(farm_token_amount, attributes);
-        reward += prev_compounded_rewards;
-
-        let initial_farming_amount =
-            self.calculate_initial_farming_amount(farm_token_amount, attributes);
+        let token_attributes = exit_farm_context
+            .farm_token
+            .attributes
+            .clone()
+            .into_part(farm_token_amount);
 
-        self.burn_farm_token_payment(&exit_farm_context.farm_token.payment);
+        let mut reward = FC::calculate_rewards(
+            self,
+            caller,
+            farm_token_amount,
+            &token_attributes,
+            &storage_cache,
+        );
+        storage_cache.reward_reserve -= &reward;
+        reward += token_attributes.get_compounded_rewards();
```

Comment:
Shouldn't the reward reserve be updated after the final reward amount, which includes the compounded rewards, is computed?

Reply:
Nope, during compound_rewards, the reward_reserve is already decremented before being added to the compounded amount.